File: faas/internal/httpx/tags_edit_test.go
package httpx

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"

	"github.com/mdk/digitaltwin2026/faas/internal/myerr"
	"github.com/mdk/digitaltwin2026/faas/internal/recordrepo"
)

// handleLogTagsAdd/Remove 纯 handler 测试（fake TagsService，校验失败/成功路径零 DB）。
func TestHandleLogTagsEdit(t *testing.T) {
	t.Parallel()
	var (
		mu   sync.Mutex
		opID string
		op   string
	)
	fakeSvc := &fakeTagsService{
		attachTag: func(_ context.Context, id, tag string) (recordrepo.EditTagsResult, *myerr.MyError) {
			mu.Lock()
			opID = id
			op = "attach"
			mu.Unlock()
			return recordrepo.EditTagsResult{From: []string{"exercise"}, To: []string{"exercise", tag}, Changed: true}, nil
		},
		detachTag: func(_ context.Context, id, tag string) (recordrepo.EditTagsResult, *myerr.MyError) {
			mu.Lock()
			opID = id
			op = "detach"
			mu.Unlock()
			return recordrepo.EditTagsResult{From: []string{"exercise", tag}, To: []string{"exercise"}, Changed: true}, nil
		},
	}
	notifier := &fakeNotifier{}
	s := &Server{TagsSvc: fakeSvc, Notifier: notifier}

	const validID = "01900000-0000-7000-8000-000000000003"

	post := func(path, body string) (*httptest.ResponseRecorder, map[string]any) {
		req := httptest.NewRequest(http.MethodPost, path, bytes.NewReader([]byte(body)))
		rr := httptest.NewRecorder()
		if path == "/api/log/tags/add" {
			s.handleLogTagsAdd(rr, req)
		} else {
			s.handleLogTagsRemove(rr, req)
		}
		var got map[string]any
		_ = json.Unmarshal(rr.Body.Bytes(), &got)
		return rr, got
	}

	t.Run("add success notifies", func(t *testing.T) {
		rr, got := post("/api/log/tags/add", `{"id":"`+validID+`","tag":"workout:arm"}`)
		if rr.Code != 200 {
			t.Fatalf("status=%d body=%s", rr.Code, rr.Body.String())
		}
		if got["success"] != true || got["id"] != validID || got["changed"] != true {
			t.Fatalf("body=%v", got)
		}
		tags := got["tags"].(map[string]any)
		join := func(v any) string {
			parts := v.([]any)
			out := make([]string, 0, len(parts))
			for _, p := range parts {
				out = append(out, p.(string))
			}
			return strings.Join(out, ",")
		}
		if join(tags["from"]) != "exercise" || join(tags["to"]) != "exercise,workout:arm" {
			t.Fatalf("tags=%v", tags)
		}
		texts := notifier.waitTexts(1)
		want := "Tags updated\nid: " + validID + "\naction: add\ntag: workout:arm\ntags: from [exercise] to [exercise, workout:arm]"
		if len(texts) != 1 || texts[0] != want {
			t.Fatalf("notify=%q want=%q", texts, want)
		}
		mu.Lock()
		defer mu.Unlock()
		if op != "attach" || opID != validID {
			t.Fatalf("op=%s id=%s", op, opID)
		}
	})

	t.Run("remove success notifies", func(t *testing.T) {
		rr, _ := post("/api/log/tags/remove", `{"id":"`+validID+`","tag":"workout:arm"}`)
		if rr.Code != 200 {
			t.Fatalf("status=%d body=%s", rr.Code, rr.Body.String())
		}
		texts := notifier.waitTexts(2)
		if len(texts) != 2 || !strings.Contains(texts[1], "action: remove") {
			t.Fatalf("notify=%q", texts)
		}
	})

	t.Run("changed false skips notify", func(t *testing.T) {
		before := len(notifier.textsLocked())
		fakeSvc.attachTag = func(_ context.Context, id, tag string) (recordrepo.EditTagsResult, *myerr.MyError) {
			return recordrepo.EditTagsResult{From: []string{"a"}, To: []string{"a"}, Changed: false}, nil
		}
		defer func() {
			fakeSvc.attachTag = nil
		}()
		rr, got := post("/api/log/tags/add", `{"id":"`+validID+`","tag":"a"}`)
		if rr.Code != 200 || got["changed"] != false {
			t.Fatalf("status=%d body=%v", rr.Code, got)
		}
		if n := len(notifier.textsLocked()); n != before {
			t.Fatalf("notify fired on changed=false: %d", n)
		}
	})
}

func TestHandleLogTagsEditValidation(t *testing.T) {
	t.Parallel()
	// 校验失败零 DB：TagsSvc 不注入（调用即 panic，可证明未进入业务层）
	s := &Server{}
	const validID = "01900000-0000-7000-8000-000000000003"

	cases := []struct {
		name, path, body, wantDetail string
		wantStatus                   int
	}{
		{"unknown key", "/api/log/tags/add", `{"id":"` + validID + `","tag":"ok","extra":1}`, `Unknown JSON key: extra`, 400},
		{"bad json", "/api/log/tags/add", `{`, "invalid JSON body", 400},
		{"invalid id", "/api/log/tags/add", `{"id":"not-a-uuid","tag":"ok"}`, "invalid record id", 400},
		{"invalid tag", "/api/log/tags/add", `{"id":"` + validID + `","tag":"bad tag"}`, "invalid tag: \"bad tag\". Tags must contain only letters, numbers, underscores, and cannot start with a number", 400},
		{"reserved tag", "/api/log/tags/add", `{"id":"` + validID + `","tag":"body:weight"}`, "tag \"body:weight\" is reserved; use the dedicated log API for this record type", 400},
		{"reserved remove", "/api/log/tags/remove", `{"id":"` + validID + `","tag":"todo:in_progress"}`, "tag \"todo:in_progress\" is reserved; use the dedicated log API for this record type", 400},
		{"not found", "/api/log/tags/add", `{"id":"` + validID + `","tag":"ok"}`, "", 404},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if c.wantDetail != "" {
				// 校验路径：不注入 service，若进业务层会 panic
				req := httptest.NewRequest(http.MethodPost, c.path, bytes.NewReader([]byte(c.body)))
				rr := httptest.NewRecorder()
				if c.path == "/api/log/tags/add" {
					s.handleLogTagsAdd(rr, req)
				} else {
					s.handleLogTagsRemove(rr, req)
				}
				if rr.Code != c.wantStatus {
					t.Fatalf("status=%d body=%s", rr.Code, rr.Body.String())
				}
				var got map[string]any
				_ = json.Unmarshal(rr.Body.Bytes(), &got)
				if got["detail"] != c.wantDetail {
					t.Fatalf("detail=%q want=%q", got["detail"], c.wantDetail)
				}
				return
			}
			// 404 路径：注入 fake 返回 not found
			s2 := &Server{TagsSvc: &fakeTagsService{attachTag: func(_ context.Context, id, tag string) (recordrepo.EditTagsResult, *myerr.MyError) {
				return recordrepo.EditTagsResult{}, myerr.NewNotFound("record " + id + " not found")
			}}}
			req := httptest.NewRequest(http.MethodPost, c.path, bytes.NewReader([]byte(c.body)))
			rr := httptest.NewRecorder()
			s2.handleLogTagsAdd(rr, req)
			if rr.Code != 404 {
				t.Fatalf("status=%d body=%s", rr.Code, rr.Body.String())
			}
			var got map[string]any
			_ = json.Unmarshal(rr.Body.Bytes(), &got)
			if got["detail"] != "record "+validID+" not found" {
				t.Fatalf("detail=%q", got["detail"])
			}
		})
	}
}

File: tests/api/tags-edit.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { NextRequest } from 'next/server'
import { POST as postTagsAdd } from '@/app/api/log/tags/add/route'
import { POST as postTagsRemove } from '@/app/api/log/tags/remove/route'
import { newNotFound } from '@/lib/myerr'

vi.mock('@/lib/tagsdb', () => ({
  tagsService: { attachTag: vi.fn(), detachTag: vi.fn() },
}))

vi.mock('@/lib/notify', () => ({
  notifyTagsEdited: vi.fn(),
  scheduleBestEffortNotify: (task: () => Promise<void>) => void task(),
}))

import { tagsService } from '@/lib/tagsdb'
import { notifyTagsEdited as notifyTagsEditedReal } from '@/lib/notify'

const attachTag = tagsService.attachTag as ReturnType<typeof vi.fn>
const detachTag = tagsService.detachTag as ReturnType<typeof vi.fn>
const notifyTagsEdited = notifyTagsEditedReal as ReturnType<typeof vi.fn>

/** 与 Go httpx tags_edit_test.go 对齐的 handler 单测（fake service，校验/成功路径零 DB） */
function rawPost(url: string, body: string): NextRequest {
  return new NextRequest(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body,
  })
}

const validID = '01900000-0000-7000-8000-000000000003'

describe('POST /api/log/tags/add', () => {
  beforeEach(() => {
    attachTag.mockReset()
    notifyTagsEdited.mockReset()
  })

  it('add success returns full shape and notifies', async () => {
    attachTag.mockResolvedValue({
      from: ['exercise'],
      to: ['exercise', 'workout:arm'],
      changed: true,
    })
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: validID, tag: 'workout:arm' })),
    )
    expect(res.status).toBe(200)
    await expect(res.json()).resolves.toEqual({
      success: true,
      id: validID,
      changed: true,
      tags: { from: ['exercise'], to: ['exercise', 'workout:arm'] },
    })
    expect(attachTag).toHaveBeenCalledWith(validID, 'workout:arm')
    expect(notifyTagsEdited).toHaveBeenCalledWith('add', validID, 'workout:arm', ['exercise'], ['exercise', 'workout:arm'])
  })

  it('duplicate returns changed:false and skips notify', async () => {
    attachTag.mockResolvedValue({ from: ['a'], to: ['a'], changed: false })
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: validID, tag: 'a' })),
    )
    expect(res.status).toBe(200)
    const body = await res.json()
    expect(body.changed).toBe(false)
    expect(notifyTagsEdited).not.toHaveBeenCalled()
  })

  it('missing record → 404 problem+json', async () => {
    attachTag.mockRejectedValue(newNotFound(`record ${validID} not found`))
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: validID, tag: 't' })),
    )
    expect(res.status).toBe(404)
    await expect(res.json()).resolves.toEqual({
      success: false,
      title: 'Not Found',
      status: 404,
      detail: `record ${validID} not found`,
    })
  })

  it('unknown key → 400', async () => {
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: validID, tag: 't', extra: 1 })),
    )
    expect(res.status).toBe(400)
    const body = await res.json()
    expect(body.detail).toBe('Unknown JSON key: extra')
    expect(attachTag).not.toHaveBeenCalled()
  })

  it('invalid id → 400', async () => {
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: 'not-a-uuid', tag: 't' })),
    )
    expect(res.status).toBe(400)
    const body = await res.json()
    expect(body.detail).toBe('invalid record id')
    expect(attachTag).not.toHaveBeenCalled()
  })

  it('invalid tag → 400', async () => {
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: validID, tag: 'bad tag' })),
    )
    expect(res.status).toBe(400)
    const body = await res.json()
    expect(body.detail).toBe(
      'invalid tag: "bad tag". Tags must contain only letters, numbers, underscores, and cannot start with a number.',
    )
    expect(attachTag).not.toHaveBeenCalled()
  })

  it('reserved tag → 400', async () => {
    const res = await postTagsAdd(
      rawPost('http://localhost/api/log/tags/add', JSON.stringify({ id: validID, tag: 'body:weight' })),
    )
    expect(res.status).toBe(400)
    const body = await res.json()
    expect(body.detail).toBe(
      'tag "body:weight" is reserved; use the dedicated log API for this record type',
    )
    expect(attachTag).not.toHaveBeenCalled()
  })

  it('malformed body → 400', async () => {
    const res = await postTagsAdd(rawPost('http://localhost/api/log/tags/add', '{not-json'))
    expect(res.status).toBe(400)
    const body = await res.json()
    expect(body.detail).toBe('invalid JSON body')
    expect(attachTag).not.toHaveBeenCalled()
  })
})

describe('POST /api/log/tags/remove', () => {
  beforeEach(() => {
    detachTag.mockReset()
    notifyTagsEdited.mockReset()
  })

  it('remove success notifies with remove action', async () => {
    detachTag.mockResolvedValue({
      from: ['exercise', 'workout:arm'],
      to: ['exercise'],
      changed: true,
    })
    const res = await postTagsRemove(
      rawPost('http://localhost/api/log/tags/remove', JSON.stringify({ id: validID, tag: 'workout:arm' })),
    )
    expect(res.status).toBe(200)
    await expect(res.json()).resolves.toEqual({
      success: true,
      id: validID,
      changed: true,
      tags: { from: ['exercise', 'workout:arm'], to: ['exercise'] },
    })
    expect(notifyTagsEdited).toHaveBeenCalledWith('remove', validID, 'workout:arm', ['exercise', 'workout:arm'], ['exercise'])
  })

  it('absent tag returns changed:false and skips notify', async () => {
    detachTag.mockResolvedValue({ from: ['a'], to: ['a'], changed: false })
    const res = await postTagsRemove(
      rawPost('http://localhost/api/log/tags/remove', JSON.stringify({ id: validID, tag: 'zzz' })),
    )
    expect(res.status).toBe(200)
    const body = await res.json()
    expect(body.changed).toBe(false)
    expect(notifyTagsEdited).not.toHaveBeenCalled()
  })

  it('reserved tag rejected symmetrically', async () => {
    const res = await postTagsRemove(
      rawPost('http://localhost/api/log/tags/remove', JSON.stringify({ id: validID, tag: 'todo:in_progress' })),
    )
    expect(res.status).toBe(400)
    const body = await res.json()
    expect(body.detail).toBe(
      'tag "todo:in_progress" is reserved; use the dedicated log API for this record type',
    )
    expect(detachTag).not.toHaveBeenCalled()
  })
})
